dali/pipeline/operator/eager_operator.h

```diff
@@ -35,42 +36,87 @@
-std::shared_ptr<TensorList<Backend>> AsTensorList(
-    const std::shared_ptr<TensorVector<Backend>> &in) {
+std::shared_ptr<TensorList<Backend>> AsTensorList(const std::shared_ptr<TensorVector<Backend>> in) {
   if (in->IsContiguous()) {
     // Filled contiguous TensorVector, we can return TensorList directly.
-    return in->AsTensorList(false);
+    auto tl = in->AsTensorList(false);
+    // Explicitly set layout (it could be empty in case of per-sample operators).
+    tl->SetLayout(in->GetLayout());
+    return tl;
   }
 
   auto tl = std::make_shared<TensorList<Backend>>();
   tl->Copy(*in);
   return tl;
 }
 
+template <typename StorageType>
+void MakeContiguous(std::shared_ptr<StorageType> storage) {}
+
+template <>
+void MakeContiguous(std::shared_ptr<TensorVector<CPUBackend>> storage) {
+  storage->SetContiguous(true);
+}
+
+template <typename Backend>
+struct Backend2Types {};
+
+template <>
+struct Backend2Types<CPUBackend> {
+  using InBackend = CPUBackend;
+  using OutBackend = CPUBackend;
+  using WSInputType = TensorVector<CPUBackend>;
+  using WSOutputType = TensorVector<CPUBackend>;
+};
+
+template <>
+struct Backend2Types<GPUBackend> {
+  using InBackend = GPUBackend;
+  using OutBackend = GPUBackend;
+  using WSInputType = TensorList<GPUBackend>;
+  using WSOutputType = TensorList<GPUBackend>;
+};
+
+template <>
+struct Backend2Types<MixedBackend> {
+  using InBackend = CPUBackend;
+  using OutBackend = GPUBackend;
+  using WSInputType = TensorVector<CPUBackend>;
+  using WSOutputType = TensorList<GPUBackend>;
+};
+
 /**
  * @brief Direct operator providing eager execution of an operator in Run.
  */
 template <typename Backend>
 class DLL_PUBLIC EagerOperator {
+  using InBackend = typename Backend2Types<Backend>::InBackend;
+  using OutBackend = typename Backend2Types<Backend>::OutBackend;
+  using WSInputType = typename Backend2Types<Backend>::WSInputType;
+  using WSOutputType = typename Backend2Types<Backend>::WSOutputType;
+
  public:
   DLL_PUBLIC inline EagerOperator(const OpSpec &spec) : EagerOperator(spec, spec.name()) {}
 
   DLL_PUBLIC inline EagerOperator(const OpSpec &spec, std::string name)
+      : EagerOperator(spec, std::move(name), shared_thread_pool->NumThreads()) {}
+
+  DLL_PUBLIC inline EagerOperator(const OpSpec &spec, std::string name, int num_threads)
       : max_batch_size_(spec.GetArgument<int>("max_batch_size")),
         op_spec_(spec),
-        name_(std::move(name)),
-        op_(InstantiateOperator(spec)) {
+        name_(std::move(name)) {
+    op_spec_.AddArg("num_threads", num_threads);
+    op_ = InstantiateOperator(op_spec_);
     num_outputs_ = op_spec_.GetSchema().CalculateOutputs(op_spec_) +
                    op_spec_.GetSchema().CalculateAdditionalOutputs(op_spec_);
   }
 
   // Runs operator using shared thread pool and shared CUDA stream.
-  template <typename InBackend, typename OutBackend>
   DLL_PUBLIC std::vector<std::shared_ptr<TensorList<OutBackend>>> Run(
```

Comment:
Does it still make sense to have those placeholders if we just define them outside of the class? This DALI_FAIL no longer would appear in any context now, unless I'm mistaken?

Reply:
done